### PR TITLE
Matrix coefficient support for H(curl) PA.

### PR DESCRIPTION
**Pull request by dylan-copeland**
Symmetric and general (asymmetric) matrix coefficients are supported for the PA mass integrator and diagonal assembly in H(curl). Unit tests verify the new capabilities. 

The implementation depends on whether the matrix coefficient is symmetric, to minimize the amount of PA data at quadrature points. This requires a way to determine whether a `MatrixCoefficient` is symmetric, and the implementation choice is debatable. The current choice is to have a flag for symmetry in `MatrixCoefficient` and a special `EvalSymmetric` function to call in the symmetric case. 

 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge| 
 | --- | --- | --- | --- | --- | --- | --- | 
| [#1620](https://github.com/mfem/mfem/pull/1620) | @dylan-copeland | @tzanio | @mlstowell + @stefanhenneking | 07/15/20 | 07/27/20 | 08/13/20 | |

### CODE REVIEW

**Comment by tzanio** [comment]
Merged in `next` for testing...

File: fem/coefficient.hpp [hunk 1]
```diff
@@ -723,6 +729,14 @@ class MatrixCoefficient
    virtual void Eval(DenseMatrix &K, ElementTransformation &T,
                      const IntegrationPoint &ip) = 0;
 
+   /** @brief Evaluate the upper triangular entries of the matrix coefficient
+       in the symmetric case, similarly to Eval. Matrix entry (i,j) is stored
+       in K[j - i + os_i] for 0 <= i <= j < width, os_0 = 0,
+       os_{i+1} = os_i + width - i. */
```
Comment (mlstowell) [inline fem/coefficient.hpp:735]: This documentation is good but perhaps something more visual could be added.  Something like:
K = {M(0,0), M(0,1), ..., M(0,w-1), M(1,1), M(1,2), ..., M(1,w-1),..., M(w-1,w-1)};

File: fem/coefficient.hpp [hunk 2]
```diff
@@ -787,10 +802,25 @@ class MatrixFunctionCoefficient : public MatrixCoefficient
       mat.SetSize(0);
    }
 
+   /// Construct a symmetric square matrix coefficient from a C-function
```
Comment (mlstowell) [inline fem/coefficient.hpp:805]: Needs a reference to `MatrixCoefficient::EvalSymmetric` or something similar to define how the vector function will be interpreted as a symmetric matrix

File: fem/bilininteg.hpp [hunk 3]
```diff
@@ -2392,6 +2392,7 @@ class VectorFEMassIntegrator: public BilinearFormIntegrator
    const DofToQuad *mapsC;         ///< Not owned. DOF-to-quad map, closed.
    const GeometricFactors *geom;   ///< Not owned
    int dim, ne, nq, dofs1D, quad1D, fetype;
+   bool symmetric = true;
```
Comment (mlstowell) [inline fem/bilininteg.hpp:2395]: Needs a comment specifying how this will be used.

**Comment by dylan-copeland** [comment]
Thanks for the suggestions, @mlstowell. Adding PA for H(curl)-H(div) mass matrices will not be too much work, and since you reviewed early I should be able to add this in time.

**Comment by mlstowell** [comment]
@dylan-copeland , no pressure.  I was thinking of a follow up PR but if you have time that would be great.

File: fem/bilininteg_vectorfe.cpp [hunk 4]
```diff
@@ -173,27 +177,63 @@ void VectorFEMassIntegrator::AssemblePA(const FiniteElementSpace &fes)
...
 
    Vector coeff(coeffDim * ne * nq);
    coeff = 1.0;
    auto coeffh = Reshape(coeff.HostWrite(), coeffDim, nq, ne);
-   if (Q || VQ)
+   if (Q || VQ || MQ)
    {
       Vector D(VQ ? coeffDim : 0);
+      DenseMatrix M(MQ ? dim : 0);
+      Vector Msymm(MQsymmDim);
```
Comment (stefanhenneking) [inline fem/bilininteg_vectorfe.cpp:197]:
```suggestion
      DenseMatrix M;
      Vector Msymm;
      if (MQ)
      {
         if (symmetric)
         {
            Msymm.SetSize(MQsymmDim)
         }
         else
         {
            M.SetSize(dim)
         }
      }
```

File: fem/bilininteg_hcurl.cpp [hunk 5]
```diff
@@ -102,20 +127,74 @@ void PAHcurlSetup3D(const int Q1D,
...
+            // First compute entries of R = MJ^{-T} = M adj(J)^T, without det J factor.
+            const double M11 = coeff(0, q, e);
+            const double M12 = coeff(1, q, e);
+            const double M13 = coeff(2, q, e);
+            const double M21 = (coeffDim == 9) ? coeff(3, q, e) : M12;
+            const double M22 = (coeffDim == 9) ? coeff(4, q, e) : coeff(3, q, e);
+            const double M23 = (coeffDim == 9) ? coeff(5, q, e) : coeff(4, q, e);
+            const double M31 = (coeffDim == 9) ? coeff(6, q, e) : M13;
+            const double M32 = (coeffDim == 9) ? coeff(7, q, e) : M23;
+            const double M33 = (coeffDim == 9) ? coeff(8, q, e) : coeff(5, q, e);
```
Comment (stefanhenneking) [inline fem/bilininteg_hcurl.cpp:142]:
```suggestion
            const double M21 = (!symmetric) ? coeff(3, q, e) : M12;
            const double M22 = (!symmetric) ? coeff(4, q, e) : coeff(3, q, e);
            const double M23 = (!symmetric) ? coeff(5, q, e) : coeff(4, q, e);
            const double M31 = (!symmetric) ? coeff(6, q, e) : M13;
            const double M32 = (!symmetric) ? coeff(7, q, e) : M23;
            const double M33 = (!symmetric) ? coeff(8, q, e) : coeff(5, q, e);
```

**Comment by dylan-copeland** [comment]
I think all the reviewers' comments have been addressed, except I will soon add PA for H(curl)-H(div) mass (not required, but it will be easier to add to this PR than to make another PR).

**Comment by tzanio** [comment]
>  except I will soon add PA for H(curl)-H(div) mass (not required, but it will be easier to add to this PR than to make another PR)

I'll wait for this then before merging in `next`

**Comment by dylan-copeland** [comment]
Mass PA is done for 3D H(curl) trial and H(div) test functions, with all possible coefficient types. Remaining tasks are H(curl) test and H(div) trial functions, diagonal assembly, and the 2D versions.

**Comment by dylan-copeland** [comment]
@mlstowell @stefanhenneking The PR is ready for review again, now that the mixed mass PA operator implementation is done. I added 2D and 3D versions of mass operators with H(curl) and H(div) as trial or test spaces, with support for all coefficient types (scalar, vector, symmetric or asymmetric matrix). There is no diagonal assembly for the mixed spaces (rectangular operators).

File: fem/bilininteg_vectorfe.cpp [hunk 6]
```diff
@@ -511,8 +536,10 @@ void VectorFEMassIntegrator::AssemblePA(const FiniteElementSpace &trial_fes,
 
    symmetric = MQ ? MQ->IsSymmetric() : true;
 
-   if (trial_fetype == mfem::FiniteElement::CURL &&
-       test_fetype == mfem::FiniteElement::DIV && dim == 3)
+   if (dim == 3 && ((trial_fetype == mfem::FiniteElement::CURL &&
+                     test_fetype == mfem::FiniteElement::DIV) ||
+                    (trial_fetype == mfem::FiniteElement::DIV &&
+                     test_fetype == mfem::FiniteElement::CURL)))
```
Comment (stefanhenneking) [inline fem/bilininteg_vectorfe.cpp]:
```suggestion
   const bool trial_curl = (trial_fetype == mfem::FiniteElement::CURL);
   const bool trial_div = (trial_fetype == mfem::FiniteElement::DIV);
   const bool test_curl = (test_fetype == mfem::FiniteElement::CURL);
   const bool test_div = (test_fetype == mfem::FiniteElement::DIV);
   if (dim == 3 && ((trial_curl && test_div) || (trial_div && test_curl)))
```

File: fem/bilininteg_vectorfe.cpp [hunk 7]
```diff
@@ -616,16 +643,22 @@ void VectorFEMassIntegrator::AssemblePA(const FiniteElementSpace &trial_fes,
       PAHdivSetup2D(quad1D, ne, ir->GetWeights(), geom->J,
                     coeff, pa_data);
    }
-   else if (trial_fetype == mfem::FiniteElement::CURL &&
-            test_fetype == mfem::FiniteElement::DIV && dim == 3 &&
+   else if (((trial_fetype == mfem::FiniteElement::CURL &&
+              test_fetype == mfem::FiniteElement::DIV) ||
+             (trial_fetype == mfem::FiniteElement::DIV &&
+              test_fetype == mfem::FiniteElement::CURL)) && dim == 3 &&
```
Comment (stefanhenneking) [inline fem/bilininteg_vectorfe.cpp]:
```suggestion
   else if (((trial_curl && test_div) || (trial_div && test_curl)) && dim == 3 &&
```

File: fem/bilininteg_vectorfe.cpp [hunk 8]
```diff
@@ -616,16 +643,22 @@ void VectorFEMassIntegrator::AssemblePA(const FiniteElementSpace &trial_fes,
...
          PAHcurlL2Setup3D(quad1D, coeffDim, ne, ir->GetWeights(),
                           coeff, pa_data);
       else
-         PAHcurlHdivSetup3D(quad1D, coeffDim, ne, ir->GetWeights(), geom->J,
-                            coeff, pa_data);
+      {
+         const bool tr = (trial_fetype == mfem::FiniteElement::DIV &&
+                          test_fetype == mfem::FiniteElement::CURL);
+         PAHcurlHdivSetup3D(quad1D, coeffDim, ne, tr, ir->GetWeights(),
+                            geom->J, coeff, pa_data);
```
Comment (stefanhenneking) [inline fem/bilininteg_vectorfe.cpp]:
```suggestion
         PAHcurlHdivSetup3D(quad1D, coeffDim, ne, (trial_div && test_curl), 
                            ir->GetWeights(), geom->J, coeff, pa_data);
```

File: fem/bilininteg_vectorfe.cpp (hunk 6 above)
Comment (stefanhenneking) [reply fem/bilininteg_vectorfe.cpp]: I think defining these additional booleans would simplify reading the remaining `if` / `else` statements. There are several additional places where these booleans could be used here that weren't modified in the commit (so I can't suggest changes in those lines).

File: fem/bilininteg_vectorfe.cpp [hunk 9]
```diff
@@ -536,10 +746,10 @@ void VectorFEMassIntegrator::AssemblePA(const FiniteElementSpace &trial_fes,
...
    symmetric = MQ ? MQ->IsSymmetric() : true;
 
-   if (dim == 3 && ((trial_fetype == mfem::FiniteElement::CURL &&
-                     test_fetype == mfem::FiniteElement::DIV) ||
-                    (trial_fetype == mfem::FiniteElement::DIV &&
-                     test_fetype == mfem::FiniteElement::CURL)))
+   if ((trial_fetype == mfem::FiniteElement::CURL &&
+        test_fetype == mfem::FiniteElement::DIV) ||
+       (trial_fetype == mfem::FiniteElement::DIV &&
+        test_fetype == mfem::FiniteElement::CURL))
```
Comment (stefanhenneking) [inline fem/bilininteg_vectorfe.cpp:752]: Similar to the other previous comment, I think these would be simpler to read using additional boolean variables.

File: fem/bilininteg_vectorfe.cpp (hunk 6 above)
Comment (dylan-copeland) [reply fem/bilininteg_vectorfe.cpp]: Thanks @stefanhenneking for the suggestions, which are now committed.

File: fem/bilininteg_vectorfe.cpp [hunk 10]
```diff
@@ -959,26 +955,21 @@ void VectorFEMassIntegrator::AddMultPA(const Vector &x, Vector &y) const
    }
    else
    {
-      if (trial_fetype == mfem::FiniteElement::CURL && test_fetype == trial_fetype)
+      if (trial_curl && test_fetype == trial_fetype)
```
Comment (stefanhenneking) [inline fem/bilininteg_vectorfe.cpp:958]:
```suggestion
      if (trial_curl && test_curl)
```

File: fem/bilininteg_vectorfe.cpp [hunk 11]
```diff
@@ -959,26 +955,21 @@ void VectorFEMassIntegrator::AddMultPA(const Vector &x, Vector &y) const
...
       {
          PAHcurlMassApply2D(dofs1D, quad1D, ne, symmetric, mapsO->B, mapsC->B,
                             mapsO->Bt, mapsC->Bt, pa_data, x, y);
       }
-      else if (trial_fetype == mfem::FiniteElement::DIV &&
-               test_fetype == trial_fetype)
+      else if (trial_div && test_fetype == trial_fetype)
```
Comment (stefanhenneking) [inline fem/bilininteg_vectorfe.cpp:963]:
```suggestion
      else if (trial_div && test_div)
```

**Comment by tzanio** [comment]
@mlstowell and @stefanhenneking can you confirm your approvals after the latest changes?

**Comment by YohannDudouit** [comment]
Could we synchronize a bit this PR with #1656 ?

**Comment by dylan-copeland** [comment]
> Could we synchronize a bit this PR with #1656 ?

@YohannDudouit The files that could have conflicts are bilininteg_hcurl.cpp and bilininteg_vectorfe.cpp. Is there a conflict you are concerned about? Since this PR will be merged to master first, you could try to pull from this branch and find the conflicts, which we can then discuss.

**Comment by YohannDudouit** [comment]
@dylan-copeland That sounds good! I'll try merging this branch to see what happens.

**Comment by tzanio** [comment]
Merged in `next` for testing...

**Comment by dylan-copeland** [comment]
@tzanio Some other PRs are waiting for this PR to get merged into master, so it would be helpful to get this merged soon.

**Pull request merged**